From 03a35fec47bd1a42b75b3373e15d0649ad562267 Mon Sep 17 00:00:00 2001
From: Guido Cella <guido@guidocella.xyz>
Date: Tue, 16 Jan 2024 19:11:18 +0100
Subject: [PATCH] allow toggling the layout before selecting a different one

---
 dwl.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dwl.c b/dwl.c
index f25ac2f..49364f8 100644
--- a/dwl.c
+++ b/dwl.c
@@ -865,7 +865,8 @@ createmon(struct wl_listener *listener, void *data)
 			m->m.y = r->y;
 			m->mfact = r->mfact;
 			m->nmaster = r->nmaster;
-			m->lt[0] = m->lt[1] = r->lt;
+			m->lt[0] = r->lt;
+			m->lt[1] = &layouts[LENGTH(layouts) > 1 && r->lt != &layouts[1]];
 			strncpy(m->ltsymbol, m->lt[m->sellt]->symbol, LENGTH(m->ltsymbol));
 			wlr_output_state_set_scale(&state, r->scale);
 			wlr_output_state_set_transform(&state, r->rr);
-- 
2.43.0

